From ac57a7f85815ea24a647e2e8dafb4ba204ec7af2 Mon Sep 17 00:00:00 2001
From: Albin B <albin.bala.krishnan@intel.com>
Date: Fri, 10 Jul 2015 22:00:12 +0530
Subject: [PATCH] [EM/USBC_PD]: Select Operating current from source cap if cap
 mismatch

When requesting the message, selecting the operating current form
the source capabilty pdo's if the voltage is mismatch. Also selecting
the previous pdo when the voltage or current is higher than the
required capabilites.

Change-Id: Idc8b3af5371d37f8d91a6191bcafc9dcb49fe248
Tracked-On: https://jira01.devtools.intel.com/browse/GMINL-13254
Signed-off-by: Albin B <albin.bala.krishnan@intel.com>
---
 drivers/usb/typec/pd/sink_port_pe.c | 23 ++++++++++-------------
 1 file changed, 10 insertions(+), 13 deletions(-)

diff --git a/drivers/usb/typec/pd/sink_port_pe.c b/drivers/usb/typec/pd/sink_port_pe.c
index 9d16008426f7..455536f6d254 100644
--- a/drivers/usb/typec/pd/sink_port_pe.c
+++ b/drivers/usb/typec/pd/sink_port_pe.c
@@ -118,6 +118,9 @@ static int snkpe_get_req_cap(struct sink_port_pe *sink,
 	rcap->cap_mismatch = true;
 
 	for (i = 0; i < num_data_obj; i++) {
+		/**
+		 * FIXME: should be selected based on the power (V*I) cap.
+		 */
 		mv = DATA_OBJ_TO_VOLT(pkt->data_obj[i]);
 		if (mv == pcap->mv) {
 			is_mv_match = true;
@@ -139,24 +142,18 @@ static int snkpe_get_req_cap(struct sink_port_pe *sink,
 		}
 	}
 
-	if (is_mv_match) {
-		rcap->obj_pos = i + 1;
-	} else {
-		/**
-		 * FIXME: should be selected based on the power (V*I) cap.
-		 * currently selecting the default vSafe5V.
-		 */
+	if (!is_mv_match) {
 		rcap->cap_mismatch = false;
-		rcap->obj_pos = 1;
+		i = 0; /* to select 1st pdo, Vsafe5V */
 	}
 
-	rcap->mv = DATA_OBJ_TO_VOLT(pkt->data_obj[rcap->obj_pos - 1]);
 	if (!rcap->cap_mismatch)
-		rcap->ma =
-			DATA_OBJ_TO_CURRENT(pkt->data_obj[rcap->obj_pos - 1]);
-	else
-		rcap->ma = pcap->ma;
+		rcap->obj_pos = i + 1; /* obj pos always starts from 1 */
+	else /* if cur is not match, select the previous pdo */
+		rcap->obj_pos = i;
 
+	rcap->mv = DATA_OBJ_TO_VOLT(pkt->data_obj[rcap->obj_pos - 1]);
+	rcap->ma = DATA_OBJ_TO_CURRENT(pkt->data_obj[rcap->obj_pos - 1]);
 
 	return 0;
 }
-- 
1.9.1

